`Start/Stop` plugin

Keyrxng opened the pull request:
Resolves https://github.com/ubiquibot/plugins-wishlist/issues/9

Keyrxng commented:
![image](https://github.com/ubiquibot/command-start-stop/assets/106303466/233b8f52-9a76-4d67-b1dc-c4220b8990bc)

gentlementlegen commented on src/types/plugin-input.ts:
```diff
@@ -0,0 +1,28 @@
+  ref: string;
+}
+
+export const startStopSchema = T.Object({
```
Should have defaults

Keyrxng replied:
Aren't the defaults passed in via the kernel, I thought they were in `generateConfiguration()`, no? I will update

0x4007 replied:
Impossible for kernel to generate defaults for every possible plugin in the future.

Keyrxng replied:
that makes sense

gentlementlegen commented on src/types/plugin-input.ts:
```diff
@@ -0,0 +1,28 @@
+}
+
+export const startStopSchema = T.Object({
+  disabledCommands: T.Array(T.String()),
```
Should have defined values for options (I guess "start", "stop")

Keyrxng replied:
I think just "start" makes sense.

It allows a user to stop a task if the command should be disabled but does not allow a new hunter to start, what do you think?

gentlementlegen replied:
It does make more sense but I don't think this would be a very common scenario. So in the end I would say let's have both there or none.

Keyrxng replied:
understood, I've added stop as default as well

gentlementlegen commented on src/types/env.ts:
```diff
@@ -0,0 +1,11 @@
+import "dotenv/config";
+
+export const envSchema = T.Object({
+  GITHUB_TOKEN: T.String(),
```
Shouldn't it match `UBIQUIBOT_TOKEN`?

gentlementlegen commented on src/plugin-config.yml:
Should not be here

gentlementlegen commented on .github/workflows/compute.yml:
Since this is supposed to be a worker maybe we don't need this part?

Keyrxng replied:
Makes sense then to also remove `main.ts` if it'll never be called in a workflow context and that is the workflow entry?

gentlementlegen replied:
The workflow entry is the file you want as long as you specify it in the configuration.

Keyrxng commented on src/handlers/shared/structured-metadata.ts:
Because it was built into it I kept it there. Remove all reference to it from this plugin or?

So the new metadata system will be a standalone plugin added to a sequence or it will be n npm package called inside any plugin? a little unsure on that one

gentlementlegen replied:
I do not know yet was it the best approach but supposedly if a plugin fail, as a GitHub Action they would dispatch an error and as a Worker they should return 500 to advise the kernel it failed. Problem if we handle that from the plugin itself that would mean every plugin should have to reimplement it.

Keyrxng replied:
As I was porting it over I was thinking that it would probably be best as a package

- callable anywhere
- useable by any plugin
- the only external call reliance is it's own calls to supabase
- best way to log more granular data without dispatching tens of workflows or fat bodied worker reqs or in the case of the kernel handling it just the fatal error reason

Keyrxng replied:
Like `rpc-handler` and `configuration` it would be part of the eco-system of plugin dev tools/libs which is starting to grow

gentlementlegen replied:
I think that:
- any fatal error should be handled by the kernel
- plugins should benefit from the new version of the `ubiquity-logged` that is capable of posting errors, and eventually can add metadata in the comment as well
- might not need a new package for this

Keyrxng commented on src/handlers/shared/stop.ts:
```diff
@@ -0,0 +1,59 @@
+import { addCommentToIssue, closePullRequestForAnIssue } from "../../utils/issue";
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export async function stop(context: Context, issue: any, sender: { id: number; login: string }, repo: any) {
```
a mistake is all, I had updated it in branched off commit and I've overlooked it starting from an earlier commit

Keyrxng commented on .cspell.json:
```diff
@@ -1,10 +1,10 @@
   "useGitignore": true,
   "language": "en",
-  "words": ["dataurl", "devpool", "outdir", "servedir"],
+  "words": ["dataurl", "devpool", "outdir", "servedir", "Supabase", "supabase", "SUPABASE", "typebox", "ubiquibot", "signoff", "sonarjs", "mswjs"],
```
You'd think so but nope those were thrown

gentlementlegen commented on src/adapters/supabase/helpers/logs.ts:
```diff
@@ -0,0 +1,387 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
```
Can't this be achieved by `unknown` or a `<T>`?

gentlementlegen commented on src/adapters/supabase/helpers/user.ts:
```diff
@@ -0,0 +1,105 @@
+const FAILED_TO_GET_USER = "Failed to get user";
+const SUCCESSFULLY_FETCHED_USER = "Successfully fetched user";
+
+interface LocationResponse {
```
I don't think this should be in the codebase

Keyrxng replied:
I removed the literals and kept it in for now, does that make more sense until the `location` stuff is fully deprecated and removed?

gentlementlegen replied:
If this is for the location why not using Supabase generated types?

Keyrxng replied:
realised it's not in use at all anymore, my mistake, removed all together as the db types are already inferred where anything `location` related is concerned here

gentlementlegen commented on src/adapters/supabase/pretty-logs.ts:
```diff
@@ -0,0 +1,190 @@
+import util from "util";
+/* eslint-disable @typescript-eslint/no-unused-vars */
```
Possible to use `unknown` or `<T>`?

gentlementlegen commented on src/types/plugin-input.ts:
```diff
@@ -0,0 +1,44 @@
+      priority: T.Array(T.String()),
+    },
+    {
+      default: {
```
Why are the labels relevant to the `start | stop`?

Keyrxng replied:
in `get-time-labels-assigned.ts` used to calc the `duration` part of the metadata and also in `analytics.ts` which is used in `get-multiplier-info.ts` which in part of the comment.

```ts
  const timeLabelsDefined = config.labels.time;
  ...
   const timeLabel = timeLabelsDefined.find((label: string) => label === _labelName);
    if (timeLabel) {
      timeLabelsAssigned.push(_label);
    }
```

gentlementlegen replied:
I get it but why should they be defined within the configuration I meant? They should just be parsed. Because the configuration from the other plugin responsible for the labels can also be changed. But you can be guaranteed to get something looking like `Time: <X DURATION` with X being a number and duration `hour | day | week`

Keyrxng replied:
I saw what you meant once I had another look at it so it's just parsing the labels from the issue itself now

rndquu commented on .github/workflows/worker-deploy.yml:
```diff
@@ -0,0 +1,29 @@
+        env:
+          SUPABASE_URL: ${{ secrets.SUPABASE_URL }}
+          SUPABASE_KEY: ${{ secrets.SUPABASE_KEY }}
+          UBIQUIBOT_TOKEN: ${{ secrets.UBIQUIBOT_TOKEN }}
```
@Keyrxng `UBIQUIBOT_TOKEN` is the github app token, right? Isn't it supposed to be passed in a plugin invocation POST request (assuming we're running the plugin as a Cloudflare worker)?

Keyrxng replied:
I'll resolve this as @gentlementlegen has answered already pretty much

rndquu commented:
@Keyrxng The plugin config example provided in the README throws `SyntaxError: "[object Object]" is not valid JSON` (I'm using [this](https://github.com/rndquu-org/test-repo/blob/159be92911c06a2585ee6088ac407b0748478dc5/.github/.ubiquibot-config.yml) one). Could you update README with a ready to use example?

Keyrxng commented:
> @Keyrxng The plugin config example provided in the README throws `SyntaxError: "[object Object]" is not valid JSON` (I'm using [this](https://github.com/rndquu-org/test-repo/blob/159be92911c06a2585ee6088ac407b0748478dc5/.github/.ubiquibot-config.yml) one). Could you update README with a ready to use example?

I probably made a mistake with it but I noticed when adjusting it that eslint was moving my comments out of the empty array.

@gentlementlegen I've brought over the template stuff, although there is no "Build" workflow, where does this come from?

gentlementlegen commented:
@Keyrxng When there is no `build` step I rather use `Knip` workflow to be done to trigger it.

gentlementlegen commented on src/types/label.ts:
```diff
@@ -0,0 +1,13 @@
+import { Static, Type } from "@sinclair/typebox";
+
+export const labelSchema = Type.Object({
```
Isn't this type provided within `Octokit`?

gentlementlegen commented on src/types/payload.ts:
I feel like a lot of these types are also directly provided by `Octokit`, any reason to create them with TypeBox?

Keyrxng replied:
I thought the same thing but also thought there must have been a reason for it, I had assumed that it had something to do with REST vs Webhook payloads or something along those lines.

I'll remove them and use the provided types since there doesn't appear to be a common knowledge reason why they were built like that

Keyrxng commented:
@gentlementlegen I noticed an error in `structuredMetadata.ts` for a regex using a char that's not supported until es2018 so that's why I've bumped that

gentlementlegen commented on src/types/payload.ts:
```diff
@@ -1,236 +1,12 @@
+export type Label = RestEndpointMethodTypes["issues"]["listLabelsOnIssue"]["response"]["data"][0];
 export type HandlerReturnValuesNoVoid = null | string | LogReturn;
+
+export const issueType = {
```
Should be an enum, and uppercase

Keyrxng replied:
I've been of the belief that enums are considered bad TS practice in most cases. I've watched/listened to quite a bit of [Matt Pocock](https://www.youtube.com/@mattpocockuk) who seems to be an authority in the TS world afaik and he mentioned once (I'm paraphrasing here) that it was asked - if they could roll things back would they introduce enums and the answer was no they would have avoided it, this was someone who was involved in the development of TS I forget their name tho.

gentlementlegen replied:
Nothing against this syntax, but The uppercase would be needed since it's used as a type imo.

Keyrxng replied:
I thought eslint would complain if we made a follow this convention my mistake

Keyrxng replied:
It's a debatable topic the whole `enum` and `const enum` thing isn't it. I was surprised to see them in use across UBQ at first actually I expected everyone to be in the against enum camp

gentlementlegen replied:
I believe it is. I do not mind either, and I don't think the overhead is that big, but if it is mentioned in best practices for modern TS it makes sense to follow these rules. Feel free to open this topic inside of the `ts-template` repo.

Keyrxng replied:
@gentlementlegen eslint does complain it just must not have updated fast enough when I checked.

> 8:14  error  Variable name `IssueType` must match one of the following formats: strictCamelCase, UPPER_CASE

This err when running `yarn format`

0x4007 replied:
> Given the context of the discussion and the specifics of TypeScript usage, here is a response and possible resolution:
> 
> 1. **Enum vs Object with `as const`:**
>    - The preference for using objects with `as const` over enums aligns with the modern TypeScript best practices. It helps keep the codebase closer to standard JavaScript and avoids the additional generated code that enums introduce. This is particularly beneficial when considering tree-shaking and overall bundle size.
>    - **Supporting Reference:** As per the TypeScript documentation, "In modern TypeScript, you may not need an enum when an object with `as const` could suffice."
> 
> 2. **Naming Convention:**
>    - The issue regarding naming conventions is valid. The ESLint rule suggests that type-like entities should follow the UPPER_CASE naming convention.
>    - To resolve this, you should update the `issueType` to `ISSUE_TYPE` to satisfy the ESLint rules without switching to an enum.
> 
> ### Updated Code
> Here's how you can update the code to adhere to the naming conventions while sticking with the object with `as const`:
> 
> ```typescript
> import { LogReturn } from "../adapters/supabase/helpers/logs";
> import { RestEndpointMethodTypes } from "@octokit/rest";
> 
> export type User = RestEndpointMethodTypes["users"]["getByUsername"]["response"]["data"];
> export type Issue = RestEndpointMethodTypes["issues"]["get"]["response"]["data"];
> export type Label = RestEndpointMethodTypes["issues"]["listLabelsOnIssue"]["response"]["data"][0];
> export type HandlerReturnValuesNoVoid = null | string | LogReturn;
> 
> export const ISSUE_TYPE = {
>   OPEN: "open",
>   CLOSED: "closed",
> } as const;
> ```
> 
> ### Summary of Points:
> - **Enum vs Object with `as const`:** Use object with `as const` to stay in line with modern TypeScript practices and keep the code closer to standard JavaScript.
> - **Naming Convention:** Change `issueType` to `ISSUE_TYPE` to comply with the ESLint naming conventions and avoid naming errors.
> 
> This approach should address both the concerns about using enums and the ESLint naming conventions.

Keyrxng replied:
So the outcome then is to use `const enum` and adjust the name yeah?

0x4007 replied:
Use the code it supplied thats fine.

0x4007 commented on src/adapters/supabase/helpers/user.ts:
```diff
@@ -0,0 +1,79 @@
+  }
+}
+
+const FAILED_TO_GET_USER = "Failed to get user";
+const SUCCESSFULLY_FETCHED_USER = "Successfully fetched user";
```
This doesn't look right to declare the const at the end of the file does this actually work

Keyrxng replied:
No it doesn't that's my mistake, it works in `.cy` files tho I think that's been my confusion here

0x4007 commented on README.md:
```diff
@@ -1,32 +1,56 @@
+- The issue is not already assigned
+- The hunter has not reached the maximum number of concurrent tasks
+- The command is not disabled at the repository or organization level
+- TODO: If the hunter meets the required karma and XP requirements
```
```suggestion
- TODO: If the hunter meets the required XP requirements
```
Don't introduce new terms. It will confuse future developers

rndquu commented:
@Keyrxng Still getting the `SyntaxError: "[object Object]" is not valid JSON` error for [this](https://github.com/rndquu-org/test-repo/blob/a06641b2d350c23809654655572caa96e416c58f/.github/.ubiquibot-config.yml) config from README file. What am I doing wrong?

Keyrxng commented:
@rndquu sorry about that you should be good to go now

gentlementlegen commented on src/adapters/supabase/helpers/user.ts:
```diff
@@ -0,0 +1,79 @@
+  issue_id: string | undefined;
+  repository_id: string | undefined;
+  node_id: string | undefined;
+  node_type: string | undefined;
```
I don't believe all of these are needed

Keyrxng replied:
you're right none of them are neither was one of the functions

gentlementlegen commented on src/utils/shared.ts:
```diff
@@ -0,0 +1,32 @@
+  const matches = label.match(/\d+/);
+  const number = matches && matches.length > 0 ? parseInt(matches[0]) || 0 : 0;
+  if (label.toLowerCase().includes("priority")) return number;
+  // throw new Error(`Label ${label} is not a priority label`);
```
```suggestion
```

0x4007 commented on src/utils/shared.ts:
```diff
@@ -0,0 +1,32 @@
+export function calculateLabelValue(label: string): number {
+  const matches = label.match(/\d+/);
+  const number = matches && matches.length > 0 ? parseInt(matches[0]) || 0 : 0;
+  if (label.toLowerCase().includes("priority")) return number;
```
We should probably handle this in a new task but these "priority" and "time" labels are intended to be customizable so these shouldn't be hardcoded strings.

gentlementlegen replied:
I actually wonder why is this needed for the `start | stop` commands? What influence do these tags have except that they have to be present on the issue in order to work? `calculateLabelValue` implies some calculation is done with these tags.

0x4007 replied:
Good point. This seems to be imported code from the v1 bot, which is later used to generate the price label. Perhaps this should be removed.

Keyrxng replied:
The below metadata is directly tied to these functions when using the `/start` command in it's current state, taken from [here](https://github.com/ubiquity/devpool-directory-bounties/issues/29#issuecomment-2176741419)

```
<!-- Ubiquity - Assignment - start - 86f94c4
{
  "duration": 3600,
  "priceLabel": {
    "id": 5905388022,
    "node_id": "LA_kwDOJhaJks8AAAABX_0R9g",
    "url": "https://api.github.com/repos/ubiquity/devpool-directory-bounties/labels/Price:%2050%20USD",
    "name": "Price: 50 USD",
    "color": "1f883d",
    "default": false,
    "description": null
  }
}
-->
```

0x4007 replied:
I think that multiplier should be applied elsewhere, like on permit generation. Seems simple enough to aggregate all rewards and slap on the multiplier on the very end. 

Reason goes with multiplier, and is read only by `/query`

However, these are not essential functions because we barely used it.

gentlementlegen commented on src/adapters/supabase/helpers/user.ts:
```diff
@@ -0,0 +1,58 @@
+      throw new Error("No wallet address found");
+    }
+
+    console.info("Successfully fetched wallet", { userId, address: data.wallets?.address });
```
All the `console` logs should be coming from the `logger` itself I believe.

Keyrxng replied:
And we should use `Logs.ts` or the simple `context` logger until the `ubiquity/logger` is refactored?

Or should it be fired up like:

```ts
context.adapters = createAdapters(supabase, context);
context.logger = new Logs(supabase, 0, "ERROR", context)
```

gentlementlegen replied:
@Keyrxng should be from the context.

gentlementlegen commented:
@rndquu That's true, but I think users cannot assign themselves without setting their wallet first, in our current workflow (bot v1)? Shall we change the workflow so users can assign themselves even without a wallet registered, and the wallet plugin would be in change to remind the user no wallet has been set?

0x4007 commented:
> @rndquu That's true, but I think users cannot assign themselves without setting their wallet first, in our current workflow (bot v1)? Shall we change the workflow so users can assign themselves even without a wallet registered, and the wallet plugin would be in change to remind the user no wallet has been set?

I believe that the old version would allow you to self assign without a registered wallet. However, we should make this configurable. Default should not allow users to self assign unless their wallet is registered!

Keyrxng commented:
Latest QA run:

- https://github.com/ubq-testing/command-start-stop/issues/1

- throws error if `startRequiresWallet` otherwise allows assignment but posts additional dedicated comment

rndquu reviewed (changes requested):
@Keyrxng Getting the following error on `yarn worker`:
```
✘ [ERROR] Could not resolve "../adapters/supabase/helpers/logs"

    src/utils/issue.ts:3:26:
      3 │ import { LogReturn } from "../adapters/supabase/helpers/logs";
```

Pls fix

gentlementlegen commented on src/types/context.ts:
```diff
@@ -25,5 +26,5 @@ export interface Context<T extends SupportedEventsU = SupportedEventsU, TU exten
   config: StartStopSettings;
   env: Env;
-  logger: Logger;
+  logger: Logger | Logs;
```
I am not sure to understand this type, why could it be `Logger | Logs`?

Keyrxng commented:
I think you all were reviewing the commit before I removed the usage of `Logs`. I've used `PrettyLogs` for the logger and removed `Logs.ts`

- https://github.com/ubq-testing/command-start-stop/issues/2 New QA

![image](https://github.com/ubiquibot/command-start-stop/assets/106303466/a48d8536-b447-47a8-aff5-3e6cb8b78c32)

rndquu commented on .github/workflows/worker-deploy.yml:
```diff
@@ -0,0 +1,27 @@
+        with:
+          wranglerVersion: "3.57.0"
+          apiToken: ${{ secrets.CLOUDFLARE_API_TOKEN }}
+          secrets: |
```
@gentlementlegen Do we need to pass `CLOUDFLARE_ACCOUNT_ID` here ([related PR](https://github.com/ubiquity/ubiquibot-kernel/pull/63))?

gentlementlegen replied:
Not required, but safer so yes it would be nice to add it here.

Keyrxng replied:
@rndquu done

rndquu commented on src/adapters/supabase/helpers/user.ts:
```diff
@@ -0,0 +1,63 @@
+    return data?.wallets?.address || null;
+  }
+
+  public async getMultiplier(userId: number, repositoryId: number) {
```
@Keyrxng As far as I understand `getMultiplier()`, `_getAccessData()` and `getLocationsFromRepo()` are not used anywhere. If so pls remove them.

Keyrxng replied:
my mistake, resolved.

rndquu reviewed (approved):
Works fine

Moved this https://github.com/ubiquibot/command-start-stop/issues/4 to a separate issue since this PR is already too big

gentlementlegen commented on .github/workflows/worker-deploy.yml:
```diff
@@ -0,0 +1,28 @@
+      - name: Setup Node
+        uses: actions/setup-node@v4
+        with:
+          node-version: "20"
```
```suggestion
          node-version: "20.10.0"
```

gentlementlegen commented on README.md:
```diff
@@ -1,32 +1,56 @@
-# `@ubiquity/ts-template`
+# `@ubiquibot/Command-Start-Stop`
```
```suggestion
# `@ubiquibot/command-start-stop`
```

gentlementlegen commented on package.json:
```diff
@@ -76,5 +83,6 @@
     ]
-  }
+  },
+  "packageManager": "yarn@4.2.2"
```
Are we really using `yarn 4` for this repo? Is there no side effect for this?

rndquu replied:
Works fine with yarn v4, at least in the "worker mode". As far I understand, [corepack](https://yarnpkg.com/corepack) (binary installed with the NodeJS) makes sure that yarn version from the `packageManager` field is used within the project.

gentlementlegen replied:
As long as it is supported within Cloudflare then we are good to go I guess!

gentlementlegen replied:
So far I had an okay experience with `bun` except for testing. `yarn@4` is fine to me as long as we don't use `Zero Install` which implies committing the whole node modules list. The main concern with yarn > 1 (which also applies to bun actually) is compatibilty with tools. For example, Vercel would accept neither (only yarn classic, npm, pnpm), so would AWS etc. Github actions don't seem to mind, neither does Cloudflare.

gentlementlegen commented on tests/main.test.ts:
```diff
@@ -65,18 +66,33 @@ describe("User start/stop", () => {
+    const issue = db.issue.findFirst({ where: { id: { equals: 6 } } }) as unknown as Issue;
+    const sender = db.users.findFirst({ where: { id: { equals: 1 } } }) as unknown as Sender;
+
+    console.log(issue.assignees, issue.assignee?.login, sender)
```
Do we want these logs?

gentlementlegen commented on tests/main.test.ts:
```diff
@@ -1,24 +1,475 @@
+    const issue = db.issue.findFirst({ where: { id: { equals: 6 } } }) as unknown as Issue;
+    const sender = db.users.findFirst({ where: { id: { equals: 1 } } }) as unknown as Sender;
+
+    console.log(issue.assignees, issue.assignee?.login, sender)
```
```suggestion
```

gentlementlegen commented on tests/main.test.ts:
```diff
@@ -1,24 +1,475 @@
+    const issue = db.issue.findFirst({ where: { id: { equals: 2 } } }) as unknown as Issue;
+    const sender = db.users.findFirst({ where: { id: { equals: 1 } } }) as unknown as Sender;
+
+    console.log(issue.assignees, issue.assignee?.login, sender)
```
```suggestion
```

0x4007 commented on README.md:
```diff
@@ -1,32 +1,56 @@
 
-To test with Cypress Studio UI, run
+- Price labels are set on the issue
+- The issue is not already assigned
```
This reminds me that it would be really interesting in the future to support teams. Where I could write on a new task 

```
/start @Keyrxng 
```

And we would both be assigned (and be eligible to split the reward.)

I know a lot of developers work as lone rangers but it could make the experience more collaborative and exciting if we have a massive tasks marketplace and teams hunt projects to co-op on together.

0x4007 commented on README.md:
```diff
@@ -1,32 +1,56 @@
+- plugin: http://localhost:4000 # or the URL where the plugin is hosted
+  name: start-stop
+  id: start-stop-command
+  description: "Allows a user to start/stop a bounty without negative XP impact"
```
```suggestion
  description: "Allows a user to start/stop a task without negative XP impact"
```

We completely stopped using "bounty" or "bounties" because this is a crypto-native term and our aspiration is to work with non crypto developers as soon as possible.

0x4007 commented on package.json:
```diff
@@ -1,7 +1,7 @@
+  "name": "@ubiquibot/command-start-stop",
   "version": "1.0.0",
-  "description": "Template repository with TypeScript support.",
+  "description": "Enables the assignment of tasks to hunters and the graceful unassignment.",
```
```suggestion
  "description": "Enables the assignment of tasks to contributors and the graceful unassignment.",
```

No more bounty hunter vernacular.

0x4007 commented on src/adapters/index.ts:
```diff
@@ -0,0 +1,14 @@
+import { Super } from "./supabase/helpers/supabase";
+import { Database } from "./supabase/types/database";
+
+export function createAdapters(supabaseClient: SupabaseClient<Database>, context: Context) {
+  return {
+    supabase: {
+      user: new User(supabaseClient, context),
+      super: new Super(supabaseClient, context),
+    },
+  };
+}
```
I think we should use each organization's own data stores. 

In exchange for having a universal/system wide registration system, we receive the following benefits:

1. decentralized data storage (within each org)
2. each org can edit the values as they see fit, providing them full control over their org and bot behaviors within. 
3. not needing to manage/maintain a supabase database (which can adversely affect the services of all of our users)
4. easier maintenance on the plugins, such as this, for not having to spin up a supabase.

Keyrxng replied:
I've branch from the latest commit of this PR and threw something together quickly but I have a question

I'm assuming there would be a dedicated `ubiquibot-plugin-storage` repo the partner would create which would be a centralized place which plugins for that install could use?

The incoming payload has the repo in which the `/start` command was used so I can't read/write to every repo and create a `start-stop-storage.json` but I could from a dedicated repo easily. 

The impl has pulled gh ids and wallets from our sb and has written to `Record<id, wallet>` then all sb logic removed and replaced with read/write/commit to repo but without the above being true, where exactly would each plugin store the data in the partner's org?

<hr/>

I think it's a good idea because any installed plugins by a partner should be trusted so each plugin having the ability to easily access the storage of other installed plugins makes things really easy for plugin DX and interoperability.

Even if the storage repo is private just knowing the storage structure of various plugins makes it easier to build on top of things. E.g. you could name the storage for this simply `wallets.json` and then other plugins need to confirm a registered user for that org checks `wallets.json`, further, `/wallet` could read/write to the same file separating logic but sharing storage

0x4007 replied:
> I've branched from the latest commit of this PR and threw something together quickly but I have a question
> 
> 
> 
> I'm assuming there would be a dedicated `ubiquibot-plugin-storage` repo the partner would create which would be a centralized place which plugins for that install could use?
> 

`.ubiquibot-config` repo 

> 
> The incoming payload has the repo in which the `/start` command was used so I can't read/write to every repo and create a `start-stop-storage.json` but I could from a dedicated repo easily. 
> 

Sorry but this doesn't make sense to me. Can you elaborate on why it's impossible to write to the event source repository?

> 
> The impl has pulled gh ids and wallets from our sb and has written to `Record<id, wallet>` then all sb logic removed and replaced with read/write/commit to repo but without the above being true, where exactly would each plugin store the data in the partner's org?
> 

Please never use acronyms in our technical discussions. 

1. It increases cognitive overhead 
2. It reduces clarity
3. It wastes time because now I need to ask you what is "impl" before I can provide you a useful response. 
4. Other contributors with less context are less likely to be able to contribute to the conversation. 

I would only accept the use of widely used acronyms that people understand without onboarding to our project. 

> 
> <hr/>
> 
> 
> 
> I think it's a good idea because any installed plugins by a partner should be trusted so each plugin having the ability to easily access the storage of other installed plugins makes things really easy for plugin DX and interoperability.
> 
> 
> 
> Even if the storage repo is private just knowing the storage structure of various plugins makes it easier to build on top of things. E.g. you could name the storage for this simply `wallets.json` and then other plugins need to confirm a registered user for that org checks `wallets.json`, further, `/wallet` could read/write to the same file separating logic but sharing storage 
> 
> 
> 
> <hr/>
> 
> 
> 
> I forget the term you used before but essentially the pre-packaged plugins we'd advise partners to install. You'd have the bare essentials like `/start|/stop`, `/wallet` etc so with that they'd each define a spot in the storage repo which could have rules like you mentioned so it can be modified or removed but it cannot be renamed and must adhere to a specific data structure, otherwise, go nuts

0x4007 commented on src/handlers/shared/check-task-stale.ts:
```diff
@@ -0,0 +1,12 @@
+export function checkTaskStale(staleTask: number, createdAt: string) {
+  let days: number | undefined;
+  let staleToDays: number | undefined;
+  let isTaskStale = false;
+
+  if (staleTask !== 0) {
+    days = Math.floor((new Date().getTime() - new Date(createdAt).getTime()) / (1000 * 60 * 60 * 24));
+    staleToDays = Math.floor(staleTask / (1000 * 60 * 60 * 24));
+    isTaskStale = days >= staleToDays;
+  }
+  return isTaskStale;
+}
```
Doesn't make sense to allow `| undefined`

Just declare it inside of the if block if thats the only place its used.

0x4007 commented on src/handlers/shared/generate-assignment-comment.ts:
```diff
@@ -0,0 +1,45 @@
+  issueCreatedAt: string,
+  issueNumber: number,
+  senderId: number,
+  duration: number | null = null
```
You could also use `0` for duration to keep it as a `number`?

0x4007 commented on src/handlers/shared/start.ts:
````diff
@@ -0,0 +1,106 @@
+  // is it assignable?
+
+  if (issue.state === ISSUE_TYPE.CLOSED) {
+    await addCommentToIssue(context, "```diff\n! The issue is closed. Please choose another unassigned bounty.\n```");
````
```suggestion
    await addCommentToIssue(context, "```diff\n! The issue is closed. Please choose another unassigned task.\n```");
```

Get rid of all "bounty" references

0x4007 commented on src/handlers/shared/stop.ts:
```diff
@@ -0,0 +1,56 @@
+  const out = { output: null };
+
+  // is it an issue?
+  if (!issue) {
+    logger.info(`Skipping '/stop' because of no issue instance`);
+    return out;
+  }
```
```suggestion
```

We can probably get rid of these checks. I know they were necessary in the old bot because of the poor type management.

0x4007 commented on src/handlers/shared/stop.ts:
`````diff
@@ -0,0 +1,56 @@
+    user: sender.login,
+  });
+
+  addCommentToIssue(context, "````diff\n+ You have been unassigned from this task.\n````").catch(logger.error);
`````
```suggestion
  addCommentToIssue(context, "```diff\n+ You have been unassigned from this task.\n```").catch(logger.error);
```

Four backticks is strange.

0x4007 commented on src/handlers/shared/structured-metadata.ts:
````diff
@@ -0,0 +1,35 @@
+  [key: string]: unknown;
+};
+
+function createStructuredMetadata<T extends object>(className: string, metadata: Metadata<T>) {
+  const jsonPretty = JSON.stringify(metadata, null, 2);
+  const stackLine = new Error().stack?.split("\n")[2] ?? "";
+  const caller = stackLine.match(/at (\S+)/)?.[1] ?? "";
+  const ubiquityMetadataHeader = `<!-- Ubiquity - ${className} - ${caller} - ${metadata.revision}`;
+
+  let metadataSerialized: string;
+  const metadataSerializedVisible = ["```json", jsonPretty, "```"].join("\n");
+  const metadataSerializedHidden = [ubiquityMetadataHeader, jsonPretty, "-->"].join("\n");
+
+  if (metadata.logMessage?.type === LogLevel.FATAL) {
+    // if the log message is fatal, then we want to show the metadata
+    metadataSerialized = [metadataSerializedVisible, metadataSerializedHidden].join("\n");
+  } else {
+    // otherwise we want to hide it
+    metadataSerialized = metadataSerializedHidden;
+  }
+
+  return metadataSerialized;
+}
+
+export default {
+  create: createStructuredMetadata,
+};
````
I wonder if we should proceed with the metadata system. 

It seems  really nice to have but we need to make it way more convenient for developers to work with (via our SDK) for this to catch on and be useful.

Keyrxng replied:
I agree it should be integrated to the SDK to be more beneficial

It would be very effective as an error wrapper/fallback of sorts. Called in catch blocks etc and before the end of plugin execution it handles posting/storage/notifying making using it a breeze

With the above actually you could hoover up metadata throughout the whole plugin flow which could be v useful (logs would do this but it would save hunting for them)

gentlementlegen replied:
Isn't this solved, at least partially, by the updates you did to `ubiquibot-logger` @Keyrxng ?

Keyrxng replied:
from a non-comment posting standpoint yes more or less this logic exists in `Logs.ts` and is returned under `metadata` in the `LogReturn` object from any Log call. It includes manipulating the stack etc but it doesn't include logic for formatting to hide content in a comment it's just a diff at the minute.

But you make a good point in that it could also just be handled by the logger which is expected to be used in all plugins

0x4007 commented on src/handlers/user-start-stop.ts:
```diff
@@ -0,0 +1,25 @@
+    throw new Error(`The '/${directive}' command is disabled for this repository.`);
+  }
+
+  if (directive === "stop") {
```
Why are these called directives? They should be called commands.

0x4007 commented on src/main.ts:
```diff
@@ -0,0 +1,48 @@
+    ref: payload.ref,
+  };
+
+  await startStopBounty(inputs, env);
```
```suggestion
  await startStopTask(inputs, env);
```

No bounty

0x4007 commented on src/main.ts:
```diff
@@ -0,0 +1,48 @@
+
+  await startStopBounty(inputs, env);
+
+  return returnDataToKernel(inputs.authToken, inputs.stateId, {});
```
```suggestion
  return toKernel(inputs.authToken, inputs.stateId, {});
```

Perhaps this is more succinct and still clear.

0x4007 commented on src/plugin.ts:
```diff
@@ -0,0 +1,31 @@
+    octokit,
+    env,
+    logger: new PrettyLogs(),
+    adapters: {} as ReturnType<typeof createAdapters>,
```
```suggestion
    adapters: createAdapters(supabase, context);
```

Doesn't this work?

Keyrxng replied:
No because `context` is not instantiated at this point and you cannot use a `this` context either

0x4007 commented on src/types/plugin-input.ts:
```diff
@@ -0,0 +1,35 @@
+  ref: string;
+}
+
+const oneDay = 24 * 60 * 60 * 1000;
```
```suggestion
const ONE_DAY = 24 * 60 * 60 * 1000;
```

0x4007 commented on src/utils/get-linked-prs.ts:
```diff
@@ -0,0 +1,44 @@
+  href: string;
+}
+
+export async function getLinkedPullRequests(context: Context, { owner, repository, issue }: GetLinkedParams): Promise<GetLinkedResults[]> {
```
I hope this works. It worked for me but @gentlementlegen had problems in the past when using it in some tests.

Keyrxng replied:
I've updated it to include one change made in #5 now checking for `pull_request` in the event. This helps restrict to only PRs and then further checks in the linked pr ensures the PR being closed is from the same repo as the current payload and belongs to the issue assignee

gentlementlegen replied:
The main issue I had was that it was not reliable as `cross-referenced` just meant quoted by another PR that may or may not close it, so ended up evaluating comments that were not relevant. Still baffled that there is not easy way to query linked PRs through the API.
I ended up parsing the body of the linked pull-request itself, but if your approach is more reliable it'd be nice to use it inside the other plugins.

Keyrxng replied:
Aye my heart was in my mouth the first time I saw it try to close this PR lol. From the QA I performed on test issues it does appear to be better yeah but only with restricting it further to match the payload repo etc

0x4007 commented on src/utils/issue.ts:
```diff
@@ -0,0 +1,171 @@
+        owner: payload.repository.owner.login,
+        repo: payload.repository.name,
+        state: ISSUE_TYPE.OPEN,
+        per_page: 1000,
```
I wonder if `1000` is the correct number. Why did you use this?

Keyrxng replied:
Must have been a typo as I know 100 is the max, surprised it didn't error but it likely defaults to max

0x4007 commented on src/utils/issue.ts:
```diff
@@ -0,0 +1,171 @@
+  }
+}
+
+//// Pull Requests \\\\
```
```suggestion
// Pull requests
```

0x4007 commented on src/utils/issue.ts:
```diff
@@ -0,0 +1,171 @@
+      state: "closed",
+    });
+  } catch (err: unknown) {
+    context.logger.fatal("Closing pull requests failed!", err);
```
I was thinking of saving fatal for emergencies that can be triaged by our team.

For example, a fatal error in the kernel occurs that can affect every partner and user of our system. So as soon as it occurs, it should ping our Telegram bot, which should relay the message in a triage group on Telegram. 

That way our incident response time can be minimal.

Obviously this should only be saved for actual, fatal emergencies. 

I don't think this qualifies.

Keyrxng replied:
So this should be a feature of the plugin logger and it should be configurable so as to ping the partner's TG and not Ubiquity TG. This will need to be well documented won't it so devs are aware fatal = tg shitstorm 😂

If a partner is running services through plugins and there is a fatal (not including the kernel because that should be ubiquity only) they should be notified in the same way?

0x4007 replied:
Our partners shouldn't be expected to be doing maintenance on our product. Obviously that's ideal but that's unrealistic to expect.

Keyrxng replied:
I agree ideal but unlikely however I meant in the sense that essentially we'd be a service provider like Cloudflare for example and if your services go down then you are notified as a service user, sometimes far later than you want to be.

How is that being addressed for partners if one of the services we provide hits a fatal for example. Manually case-by-case only if it's not fixed within a set time? Or should they also be included in this TG ping, would they have a dedicated channel we could ping maybe?

gentlementlegen replied:
`fatal` could mean incidents that we should resolve, but I think this would belong to the kernel directly since it is only a concern for us and we don't want to implement it inside every plugin.

Keyrxng replied:
Do you think there should be two loggers, one that does not include fatal and is meant for plugins while the other can have Supabase and Telegram integrations that's more for the kernel?

rndquu replied:
There are 3 entities here:
1. Kernel
2. Core ubiquity plugins
3. 3rd party plugins

Kernel and core ubiquity plugins should share the same logger instance. So if `logger.fatal()` is called in the kernel or core plugin then it's definitely an issue that we should resolve ASAP.

3rd party plugin developers should be responsible for their own plugins and maintain their own logging and monitoring solutions.

There is no need to add logger to the SDK thus allowing 3rd party plugins to use our own logging solution.

It makes sense to create a separate logger package (which is already [created](https://github.com/ubiquity/ubiquibot-logger) as far as I understand) and use it inside the kernel and core plugins.

Keyrxng replied:
The logger package was just stripped down to work with 3rd party better than with the kernel and core plugins which is why I thought two loggers made sense

So 3rd party use their own and the template should use the basic `logger.warn() => console.warn()` setup it was originally and the logger refactored for Supabase/Telegram posting

0x4007 commented on src/utils/issue.ts:
```diff
@@ -0,0 +1,171 @@
+export async function closePullRequestForAnIssue(context: Context, issueNumber: number, repository: Context["payload"]["repository"]) {
+  const logger = context.logger;
+  if (!issueNumber) {
+    throw logger.fatal("Issue is not defined");
```
Same here regarding fatal.

0x4007 commented on src/utils/issue.ts:
```diff
@@ -0,0 +1,171 @@
+  let comment = `These linked pull requests are closed: `;
+  for (let i = 0; i < linkedPullRequests.length; i++) {
+    await closePullRequest(context, linkedPullRequests[i].number);
+    comment += ` <a href="${linkedPullRequests[i].href}">#${linkedPullRequests[i].number}</a> `;
```
```suggestion
    comment += ` ${linkedPullRequests[i].href} `;
```

Not sure if this is better, but I remember seeing this render in a diff block so there's no need to dress it up if formatting isn't supported.

Keyrxng commented:
Added ubiquity logger but the exported types are not the greatest. logger needs updated to have one entry file and expose everything that's needed in it, I will do this later today

0x4007 commented on README.md:
```diff
@@ -1,32 +1,56 @@
+  command: "/(start|stop)"
+  example: "/start | /stop"
+  with:
+    disabledCommands: []
```
What is this used for

gentlementlegen commented:
@Keyrxng CSpell workflow will crash because of yarn version not defined within the workflow
```
error This project's package.json defines "packageManager": "yarn@4.2.2". However the current global version of Yarn is 1.22.22.
```

gentlementlegen commented on wrangler.toml:
```diff
@@ -0,0 +1,6 @@
+name = "ubiquibot-start-stop"
```
```suggestion
name = "ubiquibot-command-start-stop"
```

gentlementlegen commented on src/types/plugin-input.ts:
```diff
@@ -0,0 +1,35 @@
+const ONE_DAY = 24 * 60 * 60 * 1000;
+
+export const startStopSchema = T.Object({
+  isEnabled: T.Boolean(),
```
```suggestion
  isEnabled: T.Boolean({ default: true} }),
```

I think it makes more sense to enable it by default.

gentlementlegen commented:
<img width="689" alt="image" src="https://github.com/ubiquibot/command-start-stop/assets/9807008/96ca117f-55fa-4cbd-a6f0-201becdd1c10">

Just tested again an it seems there would be some formatting issues.

Second thing is a problem with the linked PRs as seen in [this test](https://github.com/Meniole/todel/issues/6#issuecomment-2212282816). This one should have been filtered out.

Otherwise the logic is working.

0x4007 commented:
> <img alt="image" width="689" src="https://private-user-images.githubusercontent.com/9807008/346307591-96ca117f-55fa-4cbd-a6f0-201becdd1c10.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA0MjIzOTksIm5iZiI6MTcyMDQyMjA5OSwicGF0aCI6Ii85ODA3MDA4LzM0NjMwNzU5MS05NmNhMTE3Zi01NWZhLTRjYmQtYTZmMC0yMDFiZWNkZDFjMTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MDhUMDcwMTM5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Njk1MGY4YjZkZTVlMjc0MzFlYTExNjE5ZjczMjZlOGEyNThlZjNlYmI5NDIwYmNjM2RmODE4ZTk0MGI1MDMwNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.s4AYHF5i8nCj46nwjBTjHsS17GWi5LfzxzWZxxHaOP8">
> Just tested again an it seems there would be some formatting issues.

This has always existed. It's fine to ship like this until we can figure a solution later (I couldn't)

> Second thing is a problem with the linked PRs as seen in [this test](https://github.com/Meniole/todel/issues/6#issuecomment-2212282816). This one should have been filtered out.
> 
> Otherwise the logic is working.

gentlementlegen commented:
@0x4007 I mentioned it because I think it worsened, if you look at our current bot
<img width="692" alt="image" src="https://github.com/ubiquibot/command-start-stop/assets/9807008/90583d80-f146-4a70-aacc-a28d7300db5f">
so I thought it'd be easier to track the root of the issue. But it can be a separate issue, surely this is not urgent.

gentlementlegen commented:
@Keyrxng Let's have those fixed in a separate PR.

gentlementlegen commented:
@Keyrxng If you could address the remaining comments so we can merge this would be nice 😄

Keyrxng commented:
Latest QA

- https://github.com/ubq-testing/command-start-stop/issues/1#issuecomment-2222953485

> Not sure if this is better, but I remember seeing this render in a diff block so there's no need to dress it up if formatting isn't supported.

It was previously rendered in a diff block and formatting is not supported. I have listed them to support formatting with your suggestion but without the diff effects. Should it be rendered in a diff block?

- https://github.com/ubq-testing/command-start-stop/issues/2#issuecomment-2222964355

0x4007 commented:
Match as close as possible to how it was and I can fix it later or something

Keyrxng commented:
@0x4007  https://github.com/ubq-testing/command-start-stop/issues/2#issuecomment-2223403247

gentlementlegen commented on src/types/context.ts:
```diff
@@ -3,7 +3,7 @@ import { Octokit } from "@octokit/rest";
 import { createAdapters } from "../adapters";
 import { Env } from "./env";
-import { Logs } from "@ubiquity-dao/ubiquibot-logger/.";
+import { Logs } from "@ubiquity-dao/ubiquibot-logger/";
```
```suggestion
import { Logs } from "@ubiquity-dao/ubiquibot-logger";
```

The slash should not be required right?

0x4007 replied:
Implies index.ts I presume

Keyrxng replied:
> Implies index.ts I presume (which is not convention)

I don't understand what you mean here not convention

> The slash should not be required right?

No It's not. One single `index.ts` files collects the important exports and makes them available via `...logger(/ | /. | /index`)

Keyrxng replied:
@0x4007 do you mean you'd rather have `/logger/logs` than just `/logger`?

Imo packages with many exports and distinct sections benefit more from have multiple entries (e.g. `permit-generation` (which needs fixed export-wise, haven't got to it yet and low priority)) 

But with only a few exports it's more convenient to have a single entry.

gentlementlegen reviewed (approved):
Please fix my last comment and good to go for me.

Keyrxng commented:
Would I merge this since everyone has approved or are we still waiting on @rndquu?

gentlementlegen commented:
@Keyrxng Should be good to go.
